File: yarl-core-java/src/test/java/hr/tjakopan/yarl/retry/RetryHandleResultTest.java
package hr.tjakopan.yarl.retry;

import hr.tjakopan.yarl.Context;
import hr.tjakopan.yarl.PolicyResult;
import hr.tjakopan.yarl.test.helpers.PolicyUtils;
import hr.tjakopan.yarl.test.helpers.TestResult;
import hr.tjakopan.yarl.test.helpers.TestResultClass;
import org.junit.Test;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.concurrent.atomic.AtomicReference;
import java.util.stream.Collectors;

import static hr.tjakopan.yarl.Functions.fromConsumer3;
import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;

public class RetryHandleResultTest {
  @Test
  public void shouldThrowWhenRetryCountIsLessThanZero() {
    assertThatThrownBy(() ->
      RetryPolicy.<TestResult>builder()
        .handleResult(TestResult.FAULT)
        .retry(-1))
      .isInstanceOf(IllegalArgumentException.class)
      .hasMessageContaining("Retry count");
  }

  @Test
  public void shouldThrowWhenOnRetryActionIsNull() {
    //noinspection ConstantConditions
    assertThatThrownBy(() ->
      RetryPolicy.<TestResult>builder()
        .handleResult(TestResult.FAULT)
        .retry(null))
      .isInstanceOf(IllegalArgumentException.class)
      .hasMessageContaining("onRetry");
  }

  @Test
  public void shouldNotReturnHandledResultWhenHandledResultRaisedSameNumberOfTimesAsRetryCount() {
    final var policy = RetryPolicy.<TestResult>builder()
      .handleResult(TestResult.FAULT)
      .retry(3);

    final var result = PolicyUtils.raiseResults(policy, TestResult.FAULT, TestResult.FAULT, TestResult.FAULT,
      TestResult.GOOD);

    assertThat(result).isEqualTo(TestResult.GOOD);
  }

  @Test
  public void shouldNotReturnHandledResultWhenOneOfTheHandledResultsRaisedSameNumberOfTimesAsRetryCount() {
    final var policy = RetryPolicy.<TestResult>builder()
      .handleResult(TestResult.FAULT)
      .handleResult(TestResult.FAULT_AGAIN)
      .retry(3);

    final var result = PolicyUtils.raiseResults(policy, TestResult.FAULT, TestResult.FAULT_AGAIN, TestResult.FAULT,
      TestResult.GOOD);

    assertThat(result).isEqualTo(TestResult.GOOD);
  }

  @Test
  public void shouldNotReturnHandledResultWhenHandledResultRaisedLessNumberOfTimesThanRetryCount() {
    final var policy = RetryPolicy.<TestResult>builder()
      .handleResult(TestResult.FAULT)
      .retry(3);

    final var result = PolicyUtils.raiseResults(policy, TestResult.FAULT, TestResult.GOOD);

    assertThat(result).isEqualTo(TestResult.GOOD);
  }

  @Test
  public void shouldNotReturnHandledResultWhenAllOfTheHandledResultsRaisedLessNumberOfTimesThanRetryCount() {
    final var policy = RetryPolicy.<TestResult>builder()
      .handleResult(TestResult.FAULT)
      .handleResult(TestResult.FAULT_AGAIN)
      .retry(3);

    final var result = PolicyUtils.raiseResults(policy, TestResult.FAULT, TestResult.FAULT_AGAIN, TestResult.GOOD);

    assertThat(result).isEqualTo(TestResult.GOOD);
  }

  @Test
  public void shouldReturnHandledResultWhenHandledResultRaisedMoreTimesThenRetryCount() {
    final var policy = RetryPolicy.<TestResult>builder()
      .handleResult(TestResult.FAULT)
      .retry(3);

    final var result = PolicyUtils.raiseResults(policy, TestResult.FAULT, TestResult.FAULT, TestResult.FAULT,
      TestResult.FAULT, TestResult.GOOD);


    assertThat(result).isEqualTo(TestResult.FAULT);
  }

  @Test
  public void shouldReturnHandledResultWhenOneOfTheHandledResultsIsRaisedMoreTimesThenRetryCount() {
    final var policy = RetryPolicy.<TestResult>builder()
      .handleResult(TestResult.FAULT)
      .handleResult(TestResult.FAULT_AGAIN)
      .retry(3);

    final var result = PolicyUtils.raiseResults(policy, TestResult.FAULT_AGAIN, TestResult.FAULT_AGAIN,
      TestResult.FAULT_AGAIN, TestResult.FAULT_AGAIN, TestResult.GOOD);

    assertThat(result).isEqualTo(TestResult.FAULT_AGAIN);
  }

  @Test
  public void shouldReturnResultWhenResultIsNotTheSpecifiedHandledResult() {
    final var policy = RetryPolicy.<TestResult>builder()
      .handleResult(TestResult.FAULT)
      .retry();

    final var result = PolicyUtils.raiseResults(policy, TestResult.FAULT_AGAIN, TestResult.GOOD);

    assertThat(result).isEqualTo(TestResult.FAULT_AGAIN);
  }

  @Test
  public void shouldReturnResultWhenResultIsNotOneOfTheSpecifiedHandledResults() {
    final var policy = RetryPolicy.<TestResult>builder()
      .handleResult(TestResult.FAULT)
      .handleResult(TestResult.FAULT_AGAIN)
      .retry();

    final var result = PolicyUtils.raiseResults(policy, TestResult.FAULT_YET_AGAIN, TestResult.GOOD);

    assertThat(result).isEqualTo(TestResult.FAULT_YET_AGAIN);
  }

  @Test
  public void shouldReturnResultWhenSpecifiedResultPredicateIsNotSatisfied() {
    final var policy = RetryPolicy.<TestResultClass>builder()
      .handleResult(r -> r.getResultCode() == TestResult.FAULT)
      .retry();

    final var result = PolicyUtils.raiseResults(policy, new TestResultClass(TestResult.FAULT_AGAIN),
      new TestResultClass(TestResult.GOOD));

    assertThat(result.getResultCode()).isEqualTo(TestResult.FAULT_AGAIN);
  }

  @Test
  public void shouldReturnResultWhenNoneOfTheSpecifiedResultPredicatesAreSatisfied() {
    final var policy = RetryPolicy.<TestResultClass>builder()
      .handleResult(r -> r.getResultCode() == TestResult.FAULT_AGAIN)
      .retry();

    final var result = PolicyUtils.raiseResults(policy, new TestResultClass(TestResult.FAULT_YET_AGAIN),
      new TestResultClass(TestResult.GOOD));

    assertThat(result.getResultCode()).isEqualTo(TestResult.FAULT_YET_AGAIN);
  }

  @Test
  public void shouldNotReturnHandledResultWhenSpecifiedResultPredicateIsSatisfied() {
    final var policy = RetryPolicy.<TestResultClass>builder()
      .handleResult(r -> r.getResultCode() == TestResult.FAULT)
      .retry();

    final var result = PolicyUtils.raiseResults(policy, new TestResultClass(TestResult.FAULT),
      new TestResultClass(TestResult.GOOD));

    assertThat(result.getResultCode()).isEqualTo(TestResult.GOOD);
  }

  @Test
  public void shouldNotReturnHandledResultWhenOneOfTheSpecifiedResultPredicatesIsSatisfied() {
    final var policy = RetryPolicy.<TestResultClass>builder()
      .handleResult(r -> r.getResultCode() == TestResult.FAULT)
      .handleResult(r -> r.getResultCode() == TestResult.FAULT_AGAIN)
      .retry();

    final var result = PolicyUtils.raiseResults(policy, new TestResultClass(TestResult.FAULT_AGAIN),
      new TestResultClass(TestResult.GOOD));

    assertThat(result.getResultCode()).isEqualTo(TestResult.GOOD);
  }

  @Test
  public void shouldCallOnRetryOnEachRetryWithTheCurrentRetryCount() {
    final var expectedRetryCounts = List.of(1, 2, 3);
    final var retryCounts = new ArrayList<Integer>(3);
    final var policy = RetryPolicy.<TestResult>builder()
      .handleResult(TestResult.FAULT)
      .retry(3, fromConsumer3(r -> (retryCount, c) -> retryCounts.add(retryCount)));

    PolicyUtils.raiseResults(policy, TestResult.FAULT, TestResult.FAULT, TestResult.FAULT, TestResult.GOOD);

    assertThat(retryCounts).containsExactlyElementsOf(expectedRetryCounts);
  }

  @Test
  public void shouldCallOnRetryOnEachRetryWithTheCurrentHandledResult() {
    final var expectedFaults = List.of("Fault #1", "Fault #2", "Fault #3");
    final var retryFaults = new ArrayList<String>(3);
    final var policy = RetryPolicy.<TestResultClass>builder()
      .handleResult(r -> r.getResultCode() == TestResult.FAULT)
      .retry(3, fromConsumer3(outcome -> (i, c) -> retryFaults.add(outcome.getOrThrow().getSomeString())));
    final var resultsToRaise = expectedFaults.stream()
      .map(s -> new TestResultClass(TestResult.FAULT, s))
      .collect(Collectors.toList());
    resultsToRaise.add(new TestResultClass(TestResult.FAULT));

    PolicyUtils.raiseResults(policy, resultsToRaise.toArray(TestResultClass[]::new));

    assertThat(retryFaults).containsExactlyElementsOf(expectedFaults);
  }

  @Test
  public void shouldNotCallOnRetryWhenNoRetriesArePerformed() {
    final var retryCalled = new AtomicBoolean(false);
    final var policy = RetryPolicy.<TestResult>builder()
      .handleResult(TestResult.FAULT)
      .retry(fromConsumer3(r -> (i, c) -> retryCalled.set(true)));

    PolicyUtils.raiseResults(policy, TestResult.GOOD);

    assertThat(retryCalled.get()).isFalse();
  }

  @Test
  public void shouldCallOnRetryWithThePassedContext() {
    final var capturedContext = new AtomicReference<Context>();
    final var policy = RetryPolicy.<TestResult>builder()
      .handleResult(TestResult.FAULT)
      .retry(fromConsumer3(r -> (i, context) -> capturedContext.set(context)));
    final var context = Context.builder()
      .contextData(new HashMap<>() {{
        put("key1", "value1");
        put("key2", "value2");
      }})
      .build();

    final var result = PolicyUtils.raiseResults(policy, context, TestResult.FAULT, TestResult.GOOD);

    assertThat(result).isEqualTo(TestResult.GOOD);
    assertThat(capturedContext.get().getContextData()).containsKeys("key1", "key2")
      .containsValues("value1", "value2");
  }

  @Test
  public void shouldCallOnRetryWithThePassedContextWhenExecuteAndCapture() {
    final var capturedContext = new AtomicReference<Context>();
    final var policy = RetryPolicy.<TestResult>builder()
      .handleResult(TestResult.FAULT)
      .retry(fromConsumer3(r -> (i, context) -> capturedContext.set(context)));
    final var context = Context.builder()
      .contextData(new HashMap<>() {{
        put("key1", "value1");
        put("key2", "value2");
      }})
      .build();

    final var result = PolicyUtils.raiseResultsOnExecuteAndCapture(policy, context, TestResult.FAULT, TestResult.GOOD);

    assertThat(result.isSuccess()).isTrue();
    //noinspection rawtypes
    assertThat(((PolicyResult.Success) result).getResult()).isEqualTo(TestResult.GOOD);
    assertThat(capturedContext.get().getContextData()).containsKeys("key1", "key2")
      .containsValues("value1", "value2");
  }

  @Test
  public void contextShouldBeEmptyIfExecuteNotCalledWithContext() {
    final var capturedContext = new AtomicReference<Context>();
    final var policy = RetryPolicy.<TestResult>builder()
      .handleResult(TestResult.FAULT)
      .retry(fromConsumer3(r -> (i, context) -> capturedContext.set(context)));

    PolicyUtils.raiseResults(policy, TestResult.FAULT, TestResult.GOOD);

    assertThat(capturedContext.get()).isNotNull();
    assertThat(capturedContext.get().getPolicyWrapKey()).isNull();
    assertThat(capturedContext.get().getPolicyKey()).isNotNull();
    assertThat(capturedContext.get().getOperationKey()).isNull();
    assertThat(capturedContext.get().getContextData()).isEmpty();
  }

  @Test
  public void shouldCreateNewContextForEachCallToExecute() {
    final var contextValue = new AtomicReference<String>();
    final var policy = RetryPolicy.<TestResult>builder()
      .handleResult(TestResult.FAULT)
      .retry(fromConsumer3(r -> (i, context) -> contextValue.set(context.getContextData().get("key").toString())));
    var context1 = Context.builder()
      .contextData(new HashMap<>() {{
        put("key", "original_value");
      }})
      .build();
    var context2 = Context.builder()
      .contextData(new HashMap<>() {{
        put("key", "new_value");
      }})
      .build();

    PolicyUtils.raiseResults(policy, context1, TestResult.FAULT, TestResult.GOOD);

    assertThat(contextValue.get()).isEqualTo("original_value");

    PolicyUtils.raiseResults(policy, context2, TestResult.FAULT, TestResult.GOOD);

    assertThat(contextValue.get()).isEqualTo("new_value");
  }

  @Test
  public void shouldCreateNewContextForEachCallToExecuteAndCapture() {
    final var contextValue = new AtomicReference<String>();
    final var policy = RetryPolicy.<TestResult>builder()
      .handleResult(TestResult.FAULT)
      .retry(fromConsumer3(r -> (i, context) -> contextValue.set(context.getContextData().get("key").toString())));
    var context1 = Context.builder()
      .contextData(new HashMap<>() {{
        put("key", "original_value");
      }})
      .build();
    var context2 = Context.builder()
      .contextData(new HashMap<>() {{
        put("key", "new_value");
      }})
      .build();

    PolicyUtils.raiseResultsOnExecuteAndCapture(policy, context1, TestResult.FAULT, TestResult.GOOD);

    assertThat(contextValue.get()).isEqualTo("original_value");

    PolicyUtils.raiseResultsOnExecuteAndCapture(policy, context2, TestResult.FAULT, TestResult.GOOD);

    assertThat(contextValue.get()).isEqualTo("new_value");
  }

  @Test
  public void shouldCreateNewStateForEachCallToPolicy() {
    final var policy = RetryPolicy.<TestResult>builder()
      .handleResult(TestResult.FAULT)
      .retry(1);

    final var result1 = PolicyUtils.raiseResults(policy, TestResult.FAULT, TestResult.GOOD);
    final var result2 = PolicyUtils.raiseResults(policy, TestResult.FAULT, TestResult.GOOD);

    assertThat(result1).isEqualTo(TestResult.GOOD);
    assertThat(result2).isEqualTo(TestResult.GOOD);
  }

  @Test
  public void shouldNotCallOnRetryWhenRetryCountIsZero() {
    final var retryInvoked = new AtomicBoolean(false);
    final var policy = RetryPolicy.<TestResult>builder()
      .handleResult(TestResult.FAULT)
      .retry(0, fromConsumer3(r -> (i, c) -> retryInvoked.set(true)));

    final var result = PolicyUtils.raiseResults(policy, TestResult.FAULT, TestResult.GOOD);

    assertThat(result).isEqualTo(TestResult.FAULT);
    assertThat(retryInvoked.get()).isFalse();
  }
}

File: yarl-core/src/test/kotlin/hr/tjakopan/yarl/retry/RetryHandleResultTest.kt
package hr.tjakopan.yarl.retry

import hr.tjakopan.yarl.Context
import hr.tjakopan.yarl.Policy
import hr.tjakopan.yarl.PolicyResult
import hr.tjakopan.yarl.test.helpers.TestResult
import hr.tjakopan.yarl.test.helpers.TestResultClass
import hr.tjakopan.yarl.test.helpers.raiseResults
import hr.tjakopan.yarl.test.helpers.raiseResultsOnExecuteAndCapture
import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.Assertions.assertThatThrownBy
import kotlin.test.Test

@Suppress("UsePropertyAccessSyntax")
class RetryHandleResultTest {
  @Test
  fun `should throw when retry count is less than zero`() {
    assertThatThrownBy {
      Policy.retry<TestResult>()
        .handleResult(TestResult.FAULT)
        .retry(-1)
    }
      .isInstanceOf(IllegalArgumentException::class.java)
      .hasMessageContaining("Retry count")
  }

  @Test
  fun `should not return handled result when handled result raised same number of times as retry count`() {
    val policy = Policy.retry<TestResult>()
      .handleResult(TestResult.FAULT)
      .retry(3)

    val result = policy.raiseResults(TestResult.FAULT, TestResult.FAULT, TestResult.FAULT, TestResult.GOOD)

    assertThat(result).isEqualTo(TestResult.GOOD)
  }

  @Test
  fun `should not return handled result when one of the handled results raised same number of times as retry count`() {
    val policy = Policy.retry<TestResult>()
      .handleResult(TestResult.FAULT)
      .handleResult(TestResult.FAULT_AGAIN)
      .retry(3)

    val result = policy.raiseResults(TestResult.FAULT, TestResult.FAULT_AGAIN, TestResult.FAULT, TestResult.GOOD)

    assertThat(result).isEqualTo(TestResult.GOOD)
  }

  @Test
  fun `should not return handled result when handled result raised less number of times than retry count`() {
    val policy = Policy.retry<TestResult>()
      .handleResult(TestResult.FAULT)
      .retry(3)

    val result = policy.raiseResults(TestResult.FAULT, TestResult.GOOD)

    assertThat(result).isEqualTo(TestResult.GOOD)
  }

  @Test
  fun `should not return handled result when all of the handled results raised less number of times than retry count`() {
    val policy = Policy.retry<TestResult>()
      .handleResult(TestResult.FAULT)
      .handleResult(TestResult.FAULT_AGAIN)
      .retry(3)

    val result = policy.raiseResults(TestResult.FAULT, TestResult.FAULT_AGAIN, TestResult.GOOD)

    assertThat(result).isEqualTo(TestResult.GOOD)
  }

  @Test
  fun `should return handled result when handled result raised more times then retry count`() {
    val policy = Policy.retry<TestResult>()
      .handleResult(TestResult.FAULT)
      .retry(3)

    val result =
      policy.raiseResults(TestResult.FAULT, TestResult.FAULT, TestResult.FAULT, TestResult.FAULT, TestResult.GOOD)

    assertThat(result).isEqualTo(TestResult.FAULT)
  }

  @Test
  fun `should return handled result when one of the handled results is raised more times then retry count`() {
    val policy = Policy.retry<TestResult>()
      .handleResult(TestResult.FAULT)
      .handleResult(TestResult.FAULT_AGAIN)
      .retry(3)

    val result = policy.raiseResults(
      TestResult.FAULT_AGAIN,
      TestResult.FAULT_AGAIN,
      TestResult.FAULT_AGAIN,
      TestResult.FAULT_AGAIN,
      TestResult.GOOD
    )

    assertThat(result).isEqualTo(TestResult.FAULT_AGAIN)
  }

  @Test
  fun `should return result when result is not the specified handled result`() {
    val policy = Policy.retry<TestResult>()
      .handleResult(TestResult.FAULT)
      .retry()

    val result = policy.raiseResults(TestResult.FAULT_AGAIN, TestResult.GOOD)

    assertThat(result).isEqualTo(TestResult.FAULT_AGAIN)
  }

  @Test
  fun `should return result when result is not one of the specified handled results`() {
    val policy = Policy.retry<TestResult>()
      .handleResult(TestResult.FAULT)
      .handleResult(TestResult.FAULT_AGAIN)
      .retry()

    val result = policy.raiseResults(TestResult.FAULT_YET_AGAIN, TestResult.GOOD)

    assertThat(result).isEqualTo(TestResult.FAULT_YET_AGAIN)
  }

  @Test
  fun `should return result when specified result predicate is not satisfied`() {
    val policy = Policy.retry<TestResultClass>()
      .handleResult { it.resultCode == TestResult.FAULT }
      .retry()

    val result = policy.raiseResults(
      TestResultClass(TestResult.FAULT_AGAIN),
      TestResultClass(TestResult.GOOD)
    )

    assertThat(result.resultCode).isEqualTo(TestResult.FAULT_AGAIN)
  }

  @Test
  fun `should return result when none of the specified result predicates are satisfied`() {
    val policy = Policy.retry<TestResultClass>()
      .handleResult { it.resultCode == TestResult.FAULT_AGAIN }
      .retry()

    val result = policy.raiseResults(
      TestResultClass(TestResult.FAULT_YET_AGAIN),
      TestResultClass(TestResult.GOOD)
    )

    assertThat(result.resultCode).isEqualTo(TestResult.FAULT_YET_AGAIN)
  }

  @Test
  fun `should not return handled result when specified result predicate is satisfied`() {
    val policy = Policy.retry<TestResultClass>()
      .handleResult { it.resultCode == TestResult.FAULT }
      .retry()

    val result = policy.raiseResults(
      TestResultClass(TestResult.FAULT),
      TestResultClass(TestResult.GOOD)
    )

    assertThat(result.resultCode).isEqualTo(TestResult.GOOD)
  }

  @Test
  fun `should not return handled result when one of the specified result predicates is satisfied`() {
    val policy = Policy.retry<TestResultClass>()
      .handleResult { it.resultCode == TestResult.FAULT }
      .handleResult { it.resultCode == TestResult.FAULT_AGAIN }
      .retry()

    val result = policy.raiseResults(
      TestResultClass(TestResult.FAULT_AGAIN),
      TestResultClass(TestResult.GOOD)
    )

    assertThat(result.resultCode).isEqualTo(TestResult.GOOD)
  }

  @Test
  fun `should call on retry on each retry with the current retry count`() {
    val expectedRetryCounts = listOf(1, 2, 3)
    val retryCounts = mutableListOf<Int>()
    val policy = Policy.retry<TestResult>()
      .handleResult(TestResult.FAULT)
      .retry(3) { _, retryCount, _ -> retryCounts.add(retryCount) }

    policy.raiseResults(TestResult.FAULT, TestResult.FAULT, TestResult.FAULT, TestResult.GOOD)

    assertThat(retryCounts).containsExactlyElementsOf(expectedRetryCounts)
  }

  @Test
  fun `should call on retry on each retry with the current handled result`() {
    val expectedFaults = listOf("Fault #1", "Fault #2", "Fault #3")
    val retryFaults = mutableListOf<String?>()
    val policy = Policy.retry<TestResultClass>()
      .handleResult { it.resultCode == TestResult.FAULT }
      .retry(3) { outcome, _, _ ->
        outcome.onSuccess { retryFaults.add(it.someString) }
      }
    val resultsToRaise = expectedFaults.map { TestResultClass(TestResult.FAULT, it) }
      .toMutableList()
    resultsToRaise.add(TestResultClass(TestResult.FAULT))

    policy.raiseResults(*resultsToRaise.toTypedArray())

    assertThat(retryFaults).containsExactlyElementsOf(expectedFaults)
  }

  @Test
  fun `should not call on retry when no retries are performed`() {
    var retryCalled = false
    val policy = Policy.retry<TestResult>()
      .handleResult(TestResult.FAULT)
      .retry { _, _, _ -> retryCalled = true }

    policy.raiseResults(TestResult.GOOD)

    assertThat(retryCalled).isFalse()
  }

  @Test
  fun `should call on retry with the passed context`() {
    var capturedContext: Context? = null
    val policy = Policy.retry<TestResult>()
      .handleResult(TestResult.FAULT)
      .retry { _, _, context -> capturedContext = context }
    val context = Context(mapOf("key1" to "value1", "key2" to "value2"))

    val result = policy.raiseResults(context, TestResult.FAULT, TestResult.GOOD)

    assertThat(result).isEqualTo(TestResult.GOOD)
    assertThat(capturedContext).containsKeys("key1", "key2")
      .containsValues("value1", "value2")
  }

  @Test
  fun `should call on retry with the passed context when execute and capture`() {
    var capturedContext: Context? = null
    val policy = Policy.retry<TestResult>()
      .handleResult(TestResult.FAULT)
      .retry { _, _, context -> capturedContext = context }
    val context = Context(mapOf("key1" to "value1", "key2" to "value2"))

    val result = policy.raiseResultsOnExecuteAndCapture(context, TestResult.FAULT, TestResult.GOOD)

    assertThat(result.isSuccess).isTrue()
    assertThat((result as PolicyResult.Success).result).isEqualTo(TestResult.GOOD)
    assertThat(capturedContext).containsKeys("key1", "key2")
      .containsValues("value1", "value2")
  }

  @Test
  fun `context should be empty if execute not called with any data`() {
    var capturedContext: Context? = null
    val policy = Policy.retry<TestResult>()
      .handleResult(TestResult.FAULT)
      .retry { _, _, context -> capturedContext = context }

    policy.raiseResults(TestResult.FAULT, TestResult.GOOD)

    assertThat(capturedContext).isEmpty()
  }

  @Test
  fun `should create new context for each call to execute`() {
    var contextValue: String? = null
    val policy = Policy.retry<TestResult>()
      .handleResult(TestResult.FAULT)
      .retry { _, _, context -> contextValue = context["key"].toString() }
    val context1 = Context(mapOf("key" to "original_value"))
    val context2 = Context(mapOf("key" to "new_value"))

    policy.raiseResults(context1, TestResult.FAULT, TestResult.GOOD)

    assertThat(contextValue).isEqualTo("original_value")

    policy.raiseResults(context2, TestResult.FAULT, TestResult.GOOD)

    assertThat(contextValue).isEqualTo("new_value")
  }

  @Test
  fun `should create new context for each call to execute and capture`() {
    var contextValue: String? = null
    val policy = Policy.retry<TestResult>()
      .handleResult(TestResult.FAULT)
      .retry { _, _, context -> contextValue = context["key"].toString() }
    val context1 = Context(mapOf("key" to "original_value"))
    val context2 = Context(mapOf("key" to "new_value"))

    policy.raiseResultsOnExecuteAndCapture(context1, TestResult.FAULT, TestResult.GOOD)

    assertThat(contextValue).isEqualTo("original_value")

    policy.raiseResultsOnExecuteAndCapture(context2, TestResult.FAULT, TestResult.GOOD)

    assertThat(contextValue).isEqualTo("new_value")
  }

  @Test
  fun `should create new state for each call to policy`() {
    val policy = Policy.retry<TestResult>()
      .handleResult(TestResult.FAULT)
      .retry(1)

    val result1 = policy.raiseResults(TestResult.FAULT, TestResult.GOOD)
    val result2 = policy.raiseResults(TestResult.FAULT, TestResult.GOOD)

    assertThat(result1).isEqualTo(TestResult.GOOD)
    assertThat(result2).isEqualTo(TestResult.GOOD)
  }

  @Test
  fun `should not call on retry when retry count is zero`() {
    var retryInvoked = false
    val policy = Policy.retry<TestResult>()
      .handleResult(TestResult.FAULT)
      .retry(0) { _, _, _ -> retryInvoked = true }

    val result = policy.raiseResults(TestResult.FAULT, TestResult.GOOD)

    assertThat(result).isEqualTo(TestResult.FAULT)
    assertThat(retryInvoked).isFalse()
  }
}
